`PathogenTrack` Quick Start Guide
--------------------

This quick start guide uses an reduced dataset generated by 10X Genomics ([SRP215370](https://www.ncbi.nlm.nih.gov/sra/?term=SRP215370)) as an example.

- [PathogenTrack quick start guide](#PathogenTrack-quick-start-guide)
  * [Step 1: Install `PathogenTrack`](#step-1-install-pathogentrack)
  * [Step 2: Download the test data](#step-2-download-the-test-data)
  * [Step 3: Excute cellranger to get scRNA count matrix](#step-3-Excute-cellranger-to-get-scRNA-count-matrix)
  * [Step 4: Run `PathogenTrack` to get the pathogens at the single-cell level](#step4-run-pathogentrack-to-get-the-pathogens-at-the-single-cell-level)

The following steps will guide you through a short tutorial of how to use the `PathogenTrack` 
package to process scRNA data and identify and quantify pathogen species at the single-cell level.
The data used comes from a gastric patient sequenced by 10X Genomics technology ([Zhang Peng, et al. 
*Cell reports* (2019): 1934-1947](https://pubmed.ncbi.nlm.nih.gov/31067475/)). We have subset the 
data to reduce the file size.

Step 1: Install `PathogenTrack`
-----------------------------------

1 . Get `PathogenTrack`.
```
git clone git@github.com:ncrna/PathogenTrack.git
```

2 . Installing `PathogenTrack` and its dependencies, users are **strongly suggested** to
use `conda` with the following command:

```
conda env create -f environment.yml
```
An environment named `PathogenTrack` will be created, and the dependencies will be installed too.
Use ```conda activate PathogenTrack``` to activate PathogenTrack environment.


3 . Prepare the host reference genome STAR index (users can change to another human genome release):
```
wget ftp://ftp.ensembl.org/pub/release-101/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.toplevel.fa.gz

gzip -d Homo_sapiens.GRCh38.dna.toplevel.fa.gz

wget ftp://ftp.ensembl.org/pub/release-101/gtf/homo_sapiens/Homo_sapiens.GRCh38.101.gtf.gz

gzip -d Homo_sapiens.GRCh38.101.gtf.gz

STAR --runThreadN 16 --runMode genomeGenerate --genomeDir ./ \
     --genomeFastaFiles ./Homo_sapiens.GRCh38.dna.toplevel.fa \
     --sjdbGTFfile ./Homo_sapiens.GRCh38.101.gtf \
     --sjdbOverhang 100
```

4 . Prepare kraken2 database (the folder is about 8 Gb):
```
wget ftp://ftp.ccb.jhu.edu/pub/data/kraken2_dbs/minikraken_8GB_202003.tgz

tar zxf minikraken_8GB_202003.tgz
```

Step 2: Prepare the test dataset
--------------------------------------

The test data comes from a gastric patient sequenced by 10X Genomics technology. It was pre-processed to
a reduced size, and you can find the test data from PathogenTrack/data directory. Then decompress these files, 
and decompress the barcodes.tsv.gz file.
```
tar zxvf PathogenTrack/data/testdata.tgz

tar zxvf PathogenTrack/data/cellranger_out.tgz

tar zxvf PathogenTrack/data/cellranger_out/barcodes.tsv.gz
```

Step 3: Excute `cellranger` to get scRNA count matrix
-----------------------------------------------------------

1 . Single-cell RNA sequencing reads are first preprocessed with single-cell quantification software 
(such as `cellranger` or `alevin`) to obtain the gene quantification matrix and cell barcode file. 
Here, we use `cellranger` as an example.
```
cd PathogenTrack/data/

cellranger count --id=test_cellranger_out \
                 --fastqs=/path/to/testdata/ \
                 --sample=test \
                 --transcriptome=/path/to/cellranger/refdata-cellranger-GRCh38-A
```

2 . Then copy the barcodes.tsv.gz file and decompress it:
```
cp test_cellranger_out/outs/filtered_feature_bc_matrix/barcodes.tsv.gz .

gzip -d barcodes.tsv.gz
```

*Note*: For tutorial, users can skip this step (step 3) by downloading the cellranger output as follows:
```
cp PathogenTrack/data/cellranger_out/barcodes.tsv.gz .

gzip -d barcodes.tsv.gz
```

Step 4: Run `PathogenTrack` to get the pathogens at the single-cell level
----------------------------------------------------------------------------

`PathogenTrack` takes raw single-cell RNAseq fastq files and the barcodes.tsv file as input. 
Users can obtain pathogen quantification matrix by running the following command:
```
python PathogenTrack.py count --project_id test \
                              --star_index /path/to/STAR/index/ \
                              --kraken_db /path/to/minikraken_8GB_20200312/ \
                              --barcode barcodes.tsv \
                              --read1 test_S1_L001_R1_001.fastq.gz \
                              --read2 test_S1_L001_R2_001.fastq.gz 
```

The full usage on how to use PathogenTrack is as follows:
```
usage: PathogenTrack.py [-h] [-v]
                        {count,trim,extract,filter,align,classify,quant} ...

positional arguments:
  {count,trim,extract,filter,align,classify,quant}
    count               one command to count microbes at single-cell levels
    trim                trim -1 at the end of barcode
    extract             extract and append barcode and UMI to the read2 name
    filter              filter out low quality / complexity reads
    align               align clean reads to the reference genome
    classify            classify unmapped reads to taxons
    quant               deduplication and quantification of microbes at
                        single-cell levels

optional arguments:
  -h, --help            show this help message and exit
  -v, --version         show program's version number and exit
```

`PathogenTrack` contains 7 functions, and users can complete the whole pathogen 
quantification process with `count` function. Otherwise, the following **6** steps can 
also complete the whole process and obtain the pathogen quantification matrix:

`trim` -> `extract` -> `filter` -> `align` -> `classify` -> `quant`

1 . Remove the trailing "-1" in the cell barcodes
```
python PathogenTrack.py trim --project_id test --barcode barcodes.tsv
```

2 . Filter and attach cell barcodes to the header of read2
```
python PathogenTrack.py extract --project_id test \
                                --read1 test_S1_L001_R1_001.fastq.gz \
                                --read2 test_S1_L001_R2_001.fastq.gz
```

3 . Filter out low-quality or low-complexity reads
```
python PathogenTrack.py filter --project_id test
```

4 . Align reads to the host reference genome and preserve the unmapped reads
```
python PathogenTrack.py align --project_id test/ --star_index /pata/to/STAR/index
```

5 . Classify the unmapped reads by kraken2
```
python PathogenTrack.py classify --project_id test/ --kraken_db /path/to/minikraken_8GB_20200312/
```

6 . Correct, deduplicate, and quantify species at the single-cell level
```
python PathogenTrack.py quant --project_id test
```

If you encountered a problem while processing your data, don't hesitate to leave an issue. We will
try our best to address the issue. Thank you for using PathogenTrack! 
